src/OrchardCore/OrchardCore.Infrastructure/Documents/DefaultDocumentSerializer.cs

```diff
@@ -57,37 +59,24 @@ internal static bool IsCompressed(byte[] data)
         return false;
     }
 
-    internal static byte[] Compress(byte[] data)
+    internal static ReadOnlySpan<byte> Compress(byte[] data)
```

Comment:
This won't work.

The memory is taken from the pool on `GetStream()` and written into by `CopyTo()`. And you correctly get the contiguous buffer with `GetBuffer().Slice()`. 

However by returning this slice to the caller, you also dispose the stream. Adn at that point the pooled buffer will be returned for another client to use it. If you are the only one on the machine then you won't see an issue. But if there is another request trying to pool some memory, they will then write in the same place you are reading the compressed data from.

The solution here is to pass the stream to this method, and the caller to use MemoryStreamFactory. -> `Compress(byte[] data, Stream destinationStream)`

Reply:
I updated the code, please check it out again